### ISSUE TITLE
AbstractPluginManager.loadPlugins fails to load plugins and returns silently if a plugin dependency was missing

### ISSUE THREAD

**Issue by slovdahl** [pf4j/pf4j#459]
Given the following scenario:
- `pluginA` and `pluginC` have no dependencies
- `pluginB` depends on `pluginA`:
   ```
   plugin-id: pluginB
   plugin-version: 1.0.0
   plugin-dependencies: pluginA
   ```
- `pluginA` is _NOT_ present in the plugin root.

When `AbstractPluginManager.loadPlugins()` is called, it returns silently and logs something like this:
 ```
2021-06-21 09:19:30 ERROR [o.p.AbstractPluginManager] :: Dependencies '[pluginB]' not found
org.pf4j.DependencyResolver$DependenciesNotFoundException: Dependencies '[pluginB]' not found
	at org.pf4j.AbstractPluginManager.resolvePlugins(AbstractPluginManager.java:794)
	at org.pf4j.AbstractPluginManager.loadPlugins(AbstractPluginManager.java:248)
```

And it also means that no other plugin is loaded (`pluginC` in this case).

Is this expected behaviour? I'm not sure what the right thing to do would be, but at least it feels like silently returning is not 100% right. Some options I could think of:
- Let the exception from the `AbstractPluginManager.resolvePlugins()` call bubble up
- Automatically detect which plugins can't be loaded and exclude them, and load the rest
- Let `AbstractPluginManager.loadPlugins()` return the plugins that could not be loaded

### Context

We call `AbstractPluginManager.loadPlugins()` during app startup, but the silent return makes it a bit challenging to detect that there was a single missing dependency, so at the moment the whole plugin system breaks down until the app is restarted if a plugin dependency was missing. We're using pf4j 3.6.0. For the time being, we can probably work around this in our own app by overriding the `loadPlugins()` and detecting the situation, but I'd be happy to submit a PR with a fix if you think it makes sense to change the behaviour.

**Comment by slovdahl** [comment]
FWIW, the same happens if there would be cyclical dependencies, the method logs and returns silently.

**Comment by slovdahl** [comment]
I now tried to override `AbstractPluginManager.loadPlugins()` and unload all plugins that have missing dependencies, but that did not work out of the box, because `AbstractPluginManager.stopPlugin(String pluginId, boolean stopDependents)` isn't able to handle `pluginState == PluginState.CREATED`. It will try to stop the plugin, but that fails with this exception:

```
java.lang.NullPointerException: null
	at org.pf4j.PluginClassLoader.loadClassFromDependencies(PluginClassLoader.java:220)
	at org.pf4j.PluginClassLoader.loadClass(PluginClassLoader.java:131)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at org.pf4j.PluginClassLoader.loadClass(PluginClassLoader.java:128)
	at org.pf4j.DefaultPluginFactory.create(DefaultPluginFactory.java:46)
	at org.pf4j.PluginWrapper.getPlugin(PluginWrapper.java:81)
	at org.pf4j.AbstractPluginManager.stopPlugin(AbstractPluginManager.java:478)
	at org.pf4j.AbstractPluginManager.unloadPlugin(AbstractPluginManager.java:284)
	at org.pf4j.AbstractPluginManager.unloadPlugin(AbstractPluginManager.java:279)
```

**Comment by slovdahl** [comment]
> Should unloadPlugin check the plugin state before calling stopPlugin?

I opened #460 with a suggested fix for this.

**Comment by slovdahl** [comment]
This turned into a rabbit hole.. :grin: 

Should `AbstractPluginManager.loadPlugin(Path pluginPath)` also unload the plugin under the hood if `resolvePlugins()` fails?

```java
    @Override
    public String loadPlugin(Path pluginPath) {
        if ((pluginPath == null) || Files.notExists(pluginPath)) {
            throw new IllegalArgumentException(String.format("Specified plugin %s does not exist!", pluginPath));
        }

        log.debug("Loading plugin from '{}'", pluginPath);

        PluginWrapper pluginWrapper = loadPluginFromPath(pluginPath);

        // try to resolve  the loaded plugin together with other possible plugins that depend on this plugin
        resolvePlugins();

        return pluginWrapper.getDescriptor().getPluginId();
    }
```

Because at the moment the plugin manager is left in some kind of strange state if `loadPlugin` fails because of missing dependencies, and cleaning up that from the caller of `loadPlugin` might not be that trivial.

**Comment by decebals** [comment]
> Is this expected behaviour? I'm not sure what the right thing to do would be, but at least it feels like silently returning is not 100% right.

The current behavior is as expected. The idea is to start the application with which plugins are OK and inform somehow the end user (UI for example) which plugins were not loaded and what was the reason. 
For this we can use [DependencyResolver.Result](https://github.com/pf4j/pf4j/blob/master/pf4j/src/main/java/org/pf4j/DependencyResolver.java#L183) or a loading report (new class) based on that `DependencyResolver.Result`.

I am sure that some users of this library (they are big companies) have implemented such a useful mechanism. I'm curious to see how they solved this

This issue is somehow connected with https://github.com/pf4j/pf4j-update/issues/22.

**Comment by slovdahl** [comment]
> The current behavior is as expected. The idea is to start the application with which plugins are OK and inform somehow the end user (UI for example) which plugins were not loaded and what was the reason.

Hm, as far as I can tell, this is not what happens with 3.6.0 though. In the situation I described above, no plugins at all are loaded if there's a single plugin with a missing dependency.

**Comment by slovdahl** [comment]
We override `AbstractPluginManager.loadPlugins()` at the moment to work around this, and this is roughly what our override does:

```diff
diff --git pf4j/src/main/java/org/pf4j/AbstractPluginManager.java pf4j/src/main/java/org/pf4j/AbstractPluginManager.java
index 8df6184..bd1b420 100644
--- pf4j/src/main/java/org/pf4j/AbstractPluginManager.java
+++ pf4j/src/main/java/org/pf4j/AbstractPluginManager.java
@@ -243,6 +243,18 @@ public abstract class AbstractPluginManager implements PluginManager {
             }
         }
 
+        // unload any plugins with missing dependencies before resolving plugins
+        List<PluginDescriptor> descriptors = new ArrayList<>();
+        for (PluginWrapper plugin : plugins.values()) {
+            descriptors.add(plugin.getDescriptor());
+        }
+
+        DependencyResolver.Result result = dependencyResolver.resolve(descriptors);
+        List<String> notFoundDependencies = result.getNotFoundDependencies();
+        for (String notFoundDependency : notFoundDependencies) {
+            unloadPlugin(notFoundDependency, true);
+        }
+
         // resolve plugins
         try {
             resolvePlugins();
```

**Comment by decebals** [comment]
@slovdahl If you are satisfied with your solution, please create a PR as a discussion point and review, maybe we can add your improvement to the core.

**Comment by slovdahl** [comment]
Sorry, never got around to submit any other PRs than #460 for the issues we found and fixed in our own subclass, so this is not completed even though it was closed with that status. @decebals I'm now working on some tests that reproduce the problem(s) to start with.

**Comment by decebals** [comment]
> Is this expected behaviour? I'm not sure what the right thing to do would be, but at least it feels like silently returning is not 100% right. Some options I could think of:
Let the exception from the AbstractPluginManager.resolvePlugins() call bubble up
Automatically detect which plugins can't be loaded and exclude them, and load the rest
Let AbstractPluginManager.loadPlugins() return the plugins that could not be loaded

I think that I prefer the variant with throw and exception as a fast fail because is easy to implement and it' reliable.
The variant with "automatically detect which plugins can't be loaded and exclude them, and load the rest" seems interesting but I'm not sure if always we want to go further only with plugins that are OK (I think you can enter an unwanted state that is hard to detect -> of course it depends on the application)
​

**Comment by decebals** [comment]
> Let the exception from the AbstractPluginManager.resolvePlugins() call bubble up

Already on master branch

> Automatically detect which plugins can't be loaded and exclude them, and load the rest

In review

@slovdahl Can you take a look and do some tests (on your application)?

**Comment by slovdahl** [comment]
For the record, #520, #563 and #564 eventually solved this, thanks @decebals ❤️ Will be available in the upcoming 3.11.0 release.
